### DISCUSSION TITLE
Access restrictions on Android 11

### DISCUSSION THREAD

**Discussion by MESepehr** [airsdk/Adobe-Runtime-Support#1173] [Q&A] [answered]
I have a local SQLite database , and copy this file from my application directory to the user document directory with these codes:

```
loader = new URLLoader();
loader.addEventListener(Event.COMPLETE, doDbCopy);
loader.dataFormat = URLLoaderDataFormat.BINARY;
loader.load(new URLRequest("Data/MWData.db"));
private function doDbCopy(e:Event):void {
            dbFile = File.documentsDirectory.resolvePath('MasterWin//db' + "//" + 'MWData.db');
            loader.removeEventListener(Event.COMPLETE, doDbCopy);
            var fileStream:FileStream = new FileStream();
            fileStream.open(dbFile, FileMode.WRITE);
            fileStream.writeBytes(loader.data);
            fileStream.close();
            connection = new SQLConnection();
            connection.addEventListener(SQLEvent.OPEN, connect)
            connection.open(dbFile, SQLMode.UPDATE);
            stat = new SQLStatement();
            stat.sqlConnection = connection;
}
```
but on android 11(API level 30) as google said :

> On Android 11 (API level 30) and higher, you cannot use the ACTION_OPEN_DOCUMENT_TREE intent action to request access to the following directories:
> 
> - The root directory of the internal storage volume.
> 
> - The root directory of each SD card volume that the device manufacturer considers to be reliable, regardless of whether the card is emulated or removable. A reliable volume is one that an app can successfully access most of the time.
> 
> - The Download directory.
> 
> Furthermore, on Android 11 (API level 30) and higher, you cannot use the ACTION_OPEN_DOCUMENT_TREE intent action to request that the user select individual files from the following directories:
> 
> - The Android/data/ directory and all subdirectories.
> 
> - The Android/obb/ directory and all subdirectories.

so as google said on this page[ Grant access to a directory's contents](https://developer.android.com/training/data-storage/shared/documents-files#grant-access-directory) do we need new permission for document directory? because when I want to copy the SQLite database on android 11 to document directory, it shows me [`Fault] exception, information=Error: Error #3001: File or directory access denied` , Is there any other solution so that the end-user does not get involved in these problems and has easy access to the database?for example I copy the database to applicationStorageDirectory but this solution did not work for me either because it showed me an error message:
`[Fault] exception, information=SQLError: 'Error #3125: Unable to open the database file.', details:'Connection closed.', operation:'open', detailID:'1001'`
FYI: I ask android.permission.WRITE_EXTERNAL_STORAGE from my users and they granted it

**Comment by ajwfrost** [comment]
I think the restriction is on the documents directory:
```
dbFile = File.documentsDirectory.resolvePath(...);
```
because the documents directory is per-user/central storage, and Android are now restricting this (but for specific file types and with particular permissions, I believe).

For this sort of usage I would think you should the application storage folder:
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/filesystem/File.html#applicationStorageDirectory
> Each AIR application has a unique, persistent application storage directory, which is created when you first access File.applicationStorageDirectory. This directory is unique to each application and user. This directory is a convenient location to store user-specific or application-specific data.

This should always be writable on Android..

**Reply by MESepehr** [reply]
thank you for your answer
but as I said before when I use
```
dbFile = File.applicationStorageDirectory.resolvepath(...)
if(dbFile.exists)
{
            connection = new SQLConnection();
            connection.addEventListener(SQLEvent.OPEN, connect)
            connection.open(dbFile, SQLMode.UPDATE);
}
```
These codes show me the following result on Android and iOS:
 `[Fault] exception, information=SQLError: 'Error #3125: Unable to open the database file.', details:'Connection closed.', operation:'open', detailID:'1001'`
While in Windows I do not see any problems and my app works like a charm!

**Reply by ajwfrost** [reply]
Ah, sorry, I missed that part at the end... we can take a look then and see what's happening under the hood.

**Reply by MESepehr** [reply]
Thankful
If you check soon, I will be grateful because many customers who have Android 11 have trouble using my application.

**Reply by ajwfrost** [reply]
Hi - this actually works for us with a couple of different devices, including Android 11; I'm wondering whether there's a difference because we're just coping a db file from the application folder to the application storage folder, whereas it looks like you're downloading this from a web location... wondering if something holds a lock on the file or similar..

Are you able to create a test package using the "apk-debug" build option, and send it to us so that we can reproduce the problem?

thanks

**Comment by MESepehr** [comment, marked as answer]
Thank you @ajwfrost, I realized that the problem was with the large size of my database(12 Mb), which was resolved with a time delay.
```
            fileStream = new FileStream();
            fileStream.open(dbFile, FileMode.WRITE);
            fileStream.writeBytes(loader.data);
            **setTimeout(completeCopyDB,4000)**// fileSteram.addEventListener(Event.Complete) doesn't work for me

            private function completeCopyDB():void
{
	    fileStream.close();
            connection = new SQLConnection();
            connection.addEventListener(SQLEvent.OPEN, connect)
            connection.open(dbFile, SQLMode.UPDATE);
            stat = new SQLStatement();
            stat.sqlConnection = connection;
}
```

**Reply by ajwfrost** [reply]
Interesting! We may need to look at this .. I know that the implementation behind the `FileStream` class is multithreaded, but I would have thought if you've opened it in synchronous mode, it should be ready for use by the database as soon as the `close` call has returned...

**Reply by marchbold** [reply]
FYI I think you need to use `openAsync()` not `open()` to get the complete event

**Reply by MESepehr** [reply]
Thank you but even with `openAsync()` I don't get the complete event!
